libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationService.java

```diff
@@ -252,9 +256,27 @@ public void onUserOffRoute(Location location, boolean userOffRoute) {
     if (userOffRoute) {
       if (location.getTime() > timeIntervalSinceLastOffRoute
         + TimeUnit.SECONDS.toMillis(mapboxNavigation.options().secondsBeforeReroute())) {
-        recentDistancesFromManeuverInMeters.clear();
-        mapboxNavigation.getEventDispatcher().onUserOffRoute(location);
         timeIntervalSinceLastOffRoute = location.getTime();
+        if (mapboxNavigation.getSessionState().lastReroutePosition() == null) {
+          recentDistancesFromManeuverInMeters.clear();
+          mapboxNavigation.getEventDispatcher().onUserOffRoute(location);
+          mapboxNavigation.setSessionState(
+            mapboxNavigation.getSessionState().toBuilder().lastReroutePosition(
+              Position.fromLngLat(location.getLongitude(), location.getLatitude())).build()
+          );
+        } else {
+          if (TurfMeasurement.distance(mapboxNavigation.getSessionState().lastReroutePosition(),
+            Position.fromLngLat(location.getLongitude(), location.getLatitude()),
+            TurfConstants.UNIT_METERS)
+            > mapboxNavigation.options().minimumDistanceBeforeRerouting()) {
+            recentDistancesFromManeuverInMeters.clear();
```

Comment:
This block is repeated above - maybe extract into its own method to prevent redundancy